docs/Migrating-Extensions-v2.0.md

```diff
@@ -0,0 +1,222 @@
+use Erusev\Parsedown\Parsedown;
+use Erusev\ParsedownExtra\Features\Abbreviations;
+
+$State = Abbreviations::from(new State);
+
+$Parsedown = new Parsedown($State);
+$actualMarkup = $Parsedown->toHtml($markdown);
+```
+
+This allows a user to have fine-grained control over which features they import, and will allow them much more control over
+combining features from multiple sources. E.g. a user may not like the way ParsedownExtra has implemented the "footnotes" feature,
+and so may wish to utilise an implementation from another source. By implementing each feature as its own `StateBearer`, we give
+users the freedom to compose features in a way that works for them.
+
+## Anatomy of the `State` Object
+
+The `State` object, generically, consists of a set of `Configurable`s. The word "set" is important here: only one instance of each
+`Configurable` may exist in a `State`. If you need to store related data in a `Configurable`, your `Configurable` needs to handle
+this containerisation itself.
+
+`State` has a special property: all `Configurable`s "exist" in any `State` object when retrieving that `Configurable` with `->get`.
+
+This means that retrieval cannot fail when using this method, though does mean that all `Configurable`s need to be "default constructable" (i.e. can be constructed into a "default" state). All `Configurable`s must therefore implement the static method
+`initial`, which must return an instance of the given `Configurable`. No initial data will be provided, but the `Configurable` **must** arrive at some sane default instance.
+
+`Configurable`s must also be immutable, unless they declare themeslves otherwise by implementing the `MutableConfigurable` interface.
+
+### Blocks
+One of the "core" `Configurable`s in Parsedown is `BlockTypes`. This contains a mapping of "markers" (a character that Parsedown
+looks for, before handing off to the block-specific parser), and a list of `Block`s that can begin parsing from a specific marker.
+Also contained, is a list of "unmarked" blocks, which Parsedown will hand off to prior to trying any marked blocks. Within marked
+blocks there is also a precedence order, where the first block type to successfully parse in this list will be the one chosen.
+
+The default value given by `BlockTypes::initial()` consists of Parsedown's default blocks. The following is a snapshot of this list:
+
+```php
+const DEFAULT_BLOCK_TYPES = [
+    '#' => [Header::class],
+    '*' => [Rule::class, TList::class],
+    '+' => [TList::class],
+    '-' => [SetextHeader::class, Table::class, Rule::class, TList::class],
+...
+```
+
+This means that if a `-` marker is found, Parsedown will first try to parse a `SetextHeader`, then try to parse a `Table`, and
+so on...
+
+A new block can be added to this list in several ways. ParsedownExtra, for example, adds a new `Abbreviation` block as follows:
+
+```php
+$BlockTypes = $State->get(BlockTypes::class)
+    ->addingMarkedLowPrecedence('*', [Abbreviation::class])
```

Comment:
Missing information about the different ways to add such as `addingMarketHighPrecedence`